src/webgpu/api/validation/resource_usages/textureUsageInRender.spec.ts

```diff
@@ -574,3 +577,68 @@ g.test('bindings_in_bundle')
       encoder.finish();
     }, !success);
   });
+
+g.test('unused_bindings_in_pipeline')
+  .params(params().combine(pbool('useBindGroup0')).combine(pbool('useBindGroup1')))
+  .fn(async t => {
+    const { useBindGroup0, useBindGroup1 } = t.params;
+    const view = t.createTexture({ usage: GPUTextureUsage.STORAGE }).createView();
+    const bindGroup0 = t.createBindGroup(0, view, 'readonly-storage-texture', 'rgba8unorm');
+    const bindGroup1 = t.createBindGroup(0, view, 'writeonly-storage-texture', 'rgba8unorm');
+
+    const wgslVertex = `
+      fn main() -> void {
+      return;
+      }
+
+      entry_point vertex = main;
+    `;
+    const binding0: string = useBindGroup0 ? '[[set 0, binding 0]] var<image> image0' : '';
+    const binding1: string = useBindGroup1 ? '[[set 1, binding 0]] var<image> image1' : '';
+    const wgslFragment =
+      binding0 +
+      binding1 +
+      `
+      fn main() -> void {
+        return;
+      }
+
+      entry_point fragment = main;
+    `;
```

Comment:
optional nit: I think this is *slightly* more readable if written more like a preprocessor?:
```suggestion
    const wgslFragment = `
      ${useBindGroup0 ? '[[set 0, binding 0]] var<image> image0;' : ''}
      ${useBindGroup1 ? '[[set 1, binding 0]] var<image> image1;' : ''}
      
      fn main() -> void { return; }
      
      entry_point fragment = main;
    `;
```

Likely we should soon consider having some very lightweight preprocessing helper utils, but I don't know what they should look like yet.

Reply:
Done. We can revisit this after preprocessing helper and 'image' can be supported in WGSL.